### ISSUE TITLE
NullPointerException when running in Headless mode

### ISSUE THREAD

**Issue by donglxd** [NationalSecurityAgency/ghidra#3435]
An error occurred when I used headless to batch import the obj files of Renesas rl78-g13 mcu to functionid, and then I tried to import a single obj file, but still reported the same error as below:
```
2021-09-18 18:13:17 INFO  (LoggingInitialization) Using log config file: jar:file:/C:/Users/Donglxd/Desktop/ghidra_10.0.3_PUBLIC_20210908/ghidra_10.0.3_PUBLIC/Ghidra/Framework/Generic/lib/Generic.jar!/generic.log4j.xml  
2021-09-18 18:13:17 INFO  (LoggingInitialization) Using log file: C:\Users\Donglxd\Desktop\rl78_import.log  
2021-09-18 18:13:17 INFO  (Preferences) Loading user preferences: C:\Users\Donglxd\.ghidra\.ghidra_10.0.3_PUBLIC\preferences  
2021-09-18 18:13:21 INFO  (ClassSearcher) Class search complete (3938 ms)  
2021-09-18 18:13:21 INFO  (SSLContextInitializer) Initializing SSL Context  
2021-09-18 18:13:21 INFO  (SecureRandomFactory) Initializing Random Number Generator...  
2021-09-18 18:13:21 INFO  (SecureRandomFactory) Random Number Generator initialization complete: SHA1PRNG  
2021-09-18 18:13:21 INFO  (ApplicationTrustManagerFactory) Trust manager disabled, cacerts have not been set  
2021-09-18 18:13:29 INFO  (HeadlessAnalyzer) HEADLESS Script Paths:
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Debug\Debugger-agent-dbgmodel-traceloader\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\Base\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Debug\Debugger\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Processors\8051\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\GnuDemangler\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Processors\DATA\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\Decompiler\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Processors\PIC\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\BytePatterns\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\MicrosoftCodeAnalyzer\ghidra_scripts
    C:\Users\Donglxd\Script
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\FunctionID\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\Python\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\FileFormats\ghidra_scripts
    C:\Users\Donglxd\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\VersionTracking\ghidra_scripts  
2021-09-18 18:13:29 INFO  (HeadlessAnalyzer) HEADLESS: execution starts  
2021-09-18 18:13:29 INFO  (HeadlessAnalyzer) Opening existing project: C:\Users\Donglxd\rl78-fidb  
2021-09-18 18:13:29 INFO  (HeadlessProject) Opening project: C:\Users\Donglxd\rl78-fidb  
2021-09-18 18:13:29 INFO  (HeadlessAnalyzer) REPORT: Processing input files:   
2021-09-18 18:13:29 INFO  (HeadlessAnalyzer)      project: C:\Users\Donglxd\rl78-fidb  
2021-09-18 18:13:29 INFO  (HeadlessAnalyzer) IMPORTING: C:\Users\Donglxd\Desktop\rl\rl78.16.test\libcom\1.0.1\19.rl.16\vsscanf.obj  
2021-09-18 18:13:30 DEBUG (QueryOpinionService) No query results found for loader Executable and Linking Format (ELF) with primary key 197 and secondary key 18177  
2021-09-18 18:13:30 ERROR (HeadlessAnalyzer) REPORT: null java.lang.NullPointerException
	at ghidra.app.util.importer.LcsHintLoadSpecChooser.choose(LcsHintLoadSpecChooser.java:62)
	at ghidra.app.util.importer.AutoImporter.getLoadSpec(AutoImporter.java:208)
	at ghidra.app.util.importer.AutoImporter.importFresh(AutoImporter.java:172)
	at ghidra.app.util.importer.AutoImporter.importFresh(AutoImporter.java:156)
	at ghidra.app.util.importer.AutoImporter.importByLookingForLcs(AutoImporter.java:91)
	at ghidra.app.util.headless.HeadlessAnalyzer.loadProgram(HeadlessAnalyzer.java:1644)
	at ghidra.app.util.headless.HeadlessAnalyzer.processFileWithImport(HeadlessAnalyzer.java:1522)
	at ghidra.app.util.headless.HeadlessAnalyzer.processWithImport(HeadlessAnalyzer.java:1688)
	at ghidra.app.util.headless.HeadlessAnalyzer.processWithImport(HeadlessAnalyzer.java:1753)
	at ghidra.app.util.headless.HeadlessAnalyzer.processLocal(HeadlessAnalyzer.java:445)
	at ghidra.app.util.headless.AnalyzeHeadless.launch(AnalyzeHeadless.java:121)
	at ghidra.GhidraLauncher.main(GhidraLauncher.java:82)
 
2021-09-18 18:13:30 ERROR (HeadlessAnalyzer) REPORT: Import failed for file: C:\Users\Donglxd\Desktop\rl\rl78.16.test\libcom\1.0.1\19.rl.16\vsscanf.obj  
```
so I checked the code at ghidra.app.util.importer.lcshintloadspecchooser.choose as follows:
```
49		@Override	
50			public LoadSpec choose(LoaderMap loaderMap) {
51			
52			// Use the highest priority loader (it will be the first one)
53			Loader loader = loaderMap.keySet().stream().findFirst().orElse(null);
54			if (loader == null) {
55			return null;
56			}
57			
58			// Try to use a known LoadSpec that matches the desired language/compiler spec
59			Collection<LoadSpec> loadSpecs = loaderMap.get(loader);
60			for (LoadSpec loadSpec : loadSpecs) {
61			LanguageCompilerSpecPair lcsPair = loadSpec.getLanguageCompilerSpec();
62	**this error happen:**		if (lcsPair.languageID.equals(languageID) &&
63			(compilerSpecID == null || lcsPair.compilerSpecID.equals(compilerSpecID))) {
64			return loadSpec;
65			}
66			}
```

so What should I do?

**Comment by donglxd** [comment]
There are also some additions:

1. I use the third-party processor module for Renesas rl78 G13（ https://github.com/xyzz/ghidra-rl78 ）, I suspect that the above errors of headless are caused by the languageid of this module

2. The window script I used to call headless. This script refers to（ https://blog.threatrack.de/2019/09/20/ghidra-fid-generator/ and https://ecos.wtf/2021/03/12/ecos-firmware-analysis-with-ghidra ）The script in is as follows:

"C:\Users\Dong\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\support\analyzeHeadless" "C:\Users\Dong" "rl78-fidb" -import "C:\Users\Dong\Desktop\rl\rl78.16.test" -recursive -scriptPath "C:\Users\Dong\Desktop\ghidra_10.0.3_PUBLIC_20210908\ghidra_10.0.3_PUBLIC\Ghidra\Features\FunctionID\ghidra_scripts" -preScript "FunctionIDHeadlessPrescript.java" -postScript "FunctionIDHeadlessPostscript.java" -processor "RL78:LE:16:default" -cspec “default”

ps:If you import a single obj, the window script command also has the same error code

3. The directory structure of the imported obj file containing rl78 is as follows:

rl
`-- rl78.16.test
    `-- libcom
        `-- 1.0.1
            `-- 19.rl.16
                |-- vsscanf.obj
                |-- vsprintf.obj
                |-- vsnprintf.obj
[...]
                |-- vscanf.obj
                |-- vprintf.obj
                `-- truncl.obj

4. The obj format is elf format. You can manually select the processor module (rl78) with the GUI of ghidra v10.0.3 and import and analyze a single obj file. However, this link is encountered during batch import using the batch import menu function of the GUI of ghidra v10.0.3（ https://github.com/NationalSecurityAgency/ghidra/issues/1934 ）The same problem that rl78 language cannot be selected

5. My ultimate goal is to generate a functionid library from obj files

**Comment by ryanmkurtz** [comment]
Can you attach one of the object files?

**Comment by donglxd** [comment]
I uploaded some obj files with this url:https://github.com/donglxd/RL78_Static_Library_To_Obj/tree/main/rl78cm4s.lib_to_obj_sampile/function2

**Comment by donglxd** [comment]
I uploaded some obj files with this url:https://github.com/donglxd/RL78_Static_Library_To_Obj/tree/main/rl78cm4s.lib_to_obj_sampile/function2

**Comment by ryanmkurtz** [comment]
In your `rl78/data/languages/` directory, try adding this `rl78.opinion` file:
```<opinions>
  <constraint loader="Executable and Linking Format (ELF)">
    <constraint compilerSpecID="default">
      <constraint primary="197" secondary="18177" processor="rl78" endian="little" size="16" variant="default"/>
    </constraint>
  </constraint>
</opinions>
```

**Comment by donglxd** [comment]
I just tested a single obj file. It seems that there is no problem. I'll try importing obj files in batches tomorrow. Thank you!

**Comment by ryanmkurtz** [comment]
Sure thing.  Let's keep this ticket open though, because we should be able to handle this case better.

**Comment by ghidra1** [comment]
A fix for the exception has been added for release with 10.0.4 which could occur when no opinion is found.

**Comment by donglxd** [comment]
I'm sorry that it took me so long to reply to you. These two days, I took the time to batch generate all the library files of rl78-g13 (CS + cc-rl compiler). According to the **rl78.option** you gave, I added some options, as follows(https://github.com/donglxd/RL78_Static_Library_To_Obj/blob/main/rl78.opinion):

<opinions>
  <constraint loader="Executable and Linking Format (ELF)">
    <constraint compilerSpecID="default">
      <constraint primary="197" secondary="18177" processor="rl78" endian="little" size="16" variant="default"/>
	  <constraint primary="197" secondary="768" processor="rl78" endian="little" size="16" variant="default"/>
	  <constraint primary="197" secondary="0" processor="rl78" endian="little" size="16" variant="default"/>
	  <constraint primary="197" secondary="776" processor="rl78" endian="little" size="16" variant="default"/>
	  <constraint primary="197" secondary="18184" processor="rl78" endian="little" size="16" variant="default"/>
	  <constraint primary="197" secondary="20232" processor="rl78" endian="little" size="16" variant="default"/>
	  <constraint primary="197" secondary="17408" processor="rl78" endian="little" size="16" variant="default"/>
    </constraint>
  </constraint>
</opinions>

Through this file, I finally generated the fidb file. I can also see the exported 2000 + functions in the tools - function ID - table viewer in the main menu. However, when I create a new analysis project to load my rl78 hex program analysis, the function name is not displayed. I searched the problem library. My situation is a little similar to this link ( https://github.com/NationalSecurityAgency/ghidra/issues/1007 ）, but I didn't understand the last responder of the post（0x6d696368). Do I need to compile the whole ghidra to get the. Fidbf file? I'm looking forward to your help to solve this problem?

a. The fidb file I generated is linked as follows:（ https://github.com/donglxd/RL78_Static_Library_To_Obj/blob/main/RL78.fidb )
b. The link to the hex file I used for the test is as follows:（https://github.com/donglxd/RL78_Static_Library_To_Obj/blob/main/TEST_FUNCTIONID_lib.hex ）I only called one isascii function in this hex


1. Screenshot of source code of hex file in CS + cc_rl compiler:
![CC_SOURCECODE](https://user-images.githubusercontent.com/62730642/134760983-8cfbb727-c5e7-4fb0-9ab7-9b11daf536ea.JPG)
2. Screenshot of the built-in disassembly code in the CS + cc_rl compiler hex file:
![CC_DISASSEMBLE](https://user-images.githubusercontent.com/62730642/134760823-b021ba01-35fd-4ba0-844f-843219dc995c.JPG)
3. Loading the hex file in ghidra does not analyze the isascii function in the functionid library
![GHIDRA_NOT_FUNCTION_LABEL](https://user-images.githubusercontent.com/62730642/134761011-e5480a1c-14c8-45f6-83bd-a6f885287abc.JPG)

**Comment by donglxd** [comment]
I just spent some time compiling ghidra (version 10.0.3) added with rl78.fidb file and generated rl78.fidbf(https://github.com/donglxd/RL78_Static_Library_To_Obj/blob/main/RL78.fidbf). The test result of Hex is disappointing or does not display isascii function. So what's the problem? I look forward to your reply!

**Comment by donglxd** [comment]
Two days have passed. Can someone help me solve the above problems?

**Comment by ryanmkurtz** [comment]
NullPointerException has been fixed by 564cab0b1bbc2e27b28256f6490eb7e17c4667c8.  Closing ticket to reflect that.

**Comment by ryanmkurtz** [comment]
Are you sure the libraries you generated the FID from were the exact ones uses in your sample?
